### PR TITLE
feat: preview dashboard filter changes

### PR DESCRIPTION
**Pull request by anirudhpillai**
## Problem
Can't add filters and preview changes to dashboard without saving the dashboard with the changes.

## Changes
When the user adds filters/changes time range the dashboard refreshes to show changes and user gets an option to 'save dashboard' with changes or 'cancel changes'.

This is also the new 'edit mode', user can press 'e' and change the layout/filters and then discard changes or save the dashboard.

https://github.com/user-attachments/assets/9f0829b6-081e-4462-a8d4-71e87f358390

## Does this work well for both Cloud and self-hosted?
N/A

## How did you test this code?
Todo: Will add E2E test once design is approved

### CODE REVIEW

File: frontend/src/scenes/dashboard/dashboardLogic.tsx
Comment (webjunkie) [inline frontend/src/scenes/dashboard/dashboardLogic.tsx:1047]: This is alright for the temporary filter update... but what now about real updates (think async query is finished and new results come in - with updated timestamps in metadata)?

**Review by webjunkie — COMMENTED** [review]
I'm wondering if the "Save" and "Cancel" is too disconnected from the filters now. Users might not see or associate it being at the top right. But I'm not sure. Maybe @Twixes has some ideas.

File: frontend/src/scenes/dashboard/dashboardLogic.tsx
Comment (anirudhpillai) [reply frontend/src/scenes/dashboard/dashboardLogic.tsx:1047]: Thanks for pointing out! I missed the attributes like `last_refresh` that get updated after the async query completes. This doesn't seem to affect most UI flows but might be needed on the insight page to show the refresh button.
 
In general, what I noticed is that even for async refresh, the metadata for the insight will already have been updated before calling `updateInsightResults` as `loadDashboard` sets the [updated dashboard](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/dashboard/dashboardLogic.tsx#L246) and then we use the [updated metadata](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/dashboard/dashboardLogic.tsx#L641) in this function

Will try to think of a way to prevent another load of the insight

Comment (webjunkie) [reply frontend/src/scenes/dashboard/dashboardLogic.tsx:1047]: > the metadata for the insight will already have been updated before calling updateInsightResults as loadDashboard sets the [updated dashboard](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/dashboard/dashboardLogic.tsx?rgh-link-date=2024-08-13T10%3A55%3A08Z#L246) and then we use the [updated metadata](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/dashboard/dashboardLogic.tsx?rgh-link-date=2024-08-13T10%3A55%3A08Z#L641) in this function

But for insights that calculate and are polling for e.g. a minute, the metadata is only then updated when the result is finished calculating. So it cannot be already in the initial load dashboard response?

File: frontend/src/scenes/dashboard/DashboardEditBar.tsx
Comment (Twixes) [inline frontend/src/scenes/dashboard/DashboardEditBar.tsx:22]: We can drop this line now that the cancel/save action is scene-level

**Review by Twixes — CHANGES_REQUESTED** [review]
This is _almost_ great – the tricky bit is that we now have _two_ dashboard editing modes, and both behave _very_ similarly: one is the already-existing layout editing mode, and now the filters editing mode. We should unify them.

**Comment by Twixes** [comment]
@webjunkie Yeah, that's good point – not necessarily obvious that changing the filter/time range activates edit mode. Some eye-catching but ephemeral indication could do a good job of alleviating that. Maybe a toast: "Now editing the dashboard – save to persist changes"?

File: frontend/src/scenes/dashboard/dashboardLogic.tsx
Comment (anirudhpillai) [reply frontend/src/scenes/dashboard/dashboardLogic.tsx]: Yes, you're right, that's what I had missed in my initial changes.
I've put the `force_cache` insight call back in for now.
Will address that in a different PR as this PR's grown much bigger now that I'm combining the filter preview with the 'edit mode' 😅

File: frontend/src/scenes/dashboard/dashboardLogic.tsx
Comment (anirudhpillai) [inline frontend/src/scenes/dashboard/dashboardLogic.tsx:383]: the kea loaders `dashboardLoading` wasn't being set to true `loadDashboard` was triggered in `afterMount`. Not sure if someone's noticed a similar issue with the loaders plugin. Can't quite figure how my change would break it especially because this reducer works as expected.

File: frontend/src/scenes/dashboard/DashboardHeader.tsx
Comment (Twixes) [inline frontend/src/scenes/dashboard/DashboardHeader.tsx:143]: We use sentence case for buttons, and not title case. [This style guide](https://posthog.com/handbook/growth/marketing/posthog-style-guide) basically applies to all text at PostHog, including in the app.

However in this case we should also follow existing UI conventions – e.g. in insight edit mode these buttons are "Cancel" and "Save". Let's follow that for consistency:
```suggestion
                                Cancel
                            </LemonButton>
                            <LemonButton
                                data-attr="dashboard-edit-mode-save"
                                type="primary"
                                onClick={() =>
                                    setDashboardMode(null, DashboardEventSource.DashboardHeaderSaveDashboard)
                                }
                                tabIndex={10}
                                disabled={dashboardLoading}
                            >
                                Save
```

File: frontend/src/scenes/dashboard/DashboardHeader.tsx
Comment (Twixes) [inline frontend/src/scenes/dashboard/DashboardHeader.tsx:132]: This does not discard changes to the dashboard's layout – they are saved immediately! That will appear like a bug to a user. It's okay to merge the current mechanics, as they allow for new possibilities worth the small chance of confusion, but a follow-up PR should also align layout editing with the Cancel/Save workflow (we can maybe put it off though – your call)

**Review by Twixes — APPROVED** [review]
Commented on a couple of rough edges, but overall looking good :shipit:

File: frontend/src/scenes/dashboard/DashboardHeader.tsx
Comment (anirudhpillai) [reply frontend/src/scenes/dashboard/DashboardHeader.tsx]: Thanks for pointing this out and updating it! Will keep that in mind. I had tried to match the edit mode wording, but matching the insight page makes a lot more sense.

File: frontend/src/scenes/dashboard/DashboardHeader.tsx
Comment (anirudhpillai) [reply frontend/src/scenes/dashboard/DashboardHeader.tsx]: This should discard changes to the layout. I disabled saving the layout changes instantly from [here](https://github.com/PostHog/posthog/pull/24303/files#diff-749ecf967fe174bf1f31fd4d01262d1335dda1ad07cf98b5f618e936e06ff781L919)
In my updated [PR desc video](https://github.com/PostHog/posthog/pull/24303#issue-2459761010), I show how the layout changes reset when you click 'Cancel'. But there might be a bug if it still doesn't work for you? 😓

File: posthog/utils.py
Comment (Twixes) [inline posthog/utils.py:1048]: This might result in a JSON parsing error – in that case we should catch it and throw a ValidationError instead to avoid a 500

File: posthog/api/dashboards/dashboard.py
Comment (Twixes) [inline posthog/api/dashboards/dashboard.py:294]: This seems wasteful, because it means this method runs `UPDATE posthog_dashboard ...` twice – can we merge this operation with the existing `super().update(instance, validated_data)`?

File: frontend/src/scenes/dashboard/dashboardLogic.tsx
Comment (Twixes) [inline frontend/src/scenes/dashboard/dashboardLogic.tsx:141]: Let's use camelCase in JS:
```suggestion
    filtersOverride?: DashboardFilter
```

**Review by Twixes — APPROVED** [review]
Oh, yes, you're right @anirudhpillai – don't know what happened, but working as expected now. That's great.

When I commented before I only took a cursory look at the and focused on the experience. Now I went through the code in a bit more detail – left three comments on quick improvements we should do before merging. Overall good to go. Nice job

**Pull request merged**
